Refactor/frontend backend split

ThanhDatVu111 opened the pull request:
## 🚀 What does this PR do?

> A short summary of what’s changed and why.
Refactor the code base to separate frontend and backend.

This is why the PR has around 100 commits — please don't panic when reviewing 😄.

**Changes include:**
    
    - Move frontend code into frontend/ folder with src/ subdirectory
    
    - Move backend (Supabase config, migrations) into backend/ folder
    
    - Configure NativeWind and TailwindCSS properly inside the frontend
    
    - Fix Metro bundler config (metro.config.js) and Babel config (babel.config.js)
    
    - Correct the Expo entry point to src/app/index.tsx
    
    - Fix font loading paths and asset structure
    
    - Set up Prettier formatting inside the frontend
    
    - Install missing dependencies (react-native-reanimated, etc.)
    
    - Cleanup unnecessary files at the root level

---

## ✅ Checklist

- [x] Code compiles & runs
- [x] Lint checks pass
- [x] Updated `.env.example` if needed
- [x] PR linked to related issue/ticket (if applicable)

---

## 🔍 How to test this

> Describe how reviewers can test this — steps, screenshots, etc.
cd frontend

    npm install
    
    npm start or npx expo start
    
    App should load successfully showing the correct entry page.
    
    Tailwind classes (e.g., background colors, text sizing) should work as expected.
    
    Fonts like Poppins-Bold and images like logo and polaroid should load properly.
    
    Backend (Supabase folder) remains untouched and separated under /backend.

---

## 🧠 Extra Notes

> Any decisions made? Tradeoffs? Gotchas?

- This refactor is necessary to separate responsibilities between frontend and backend moving forward.

- Expo, NativeWind, and Supabase are now properly organized for easier development and maintenance.

- Future updates will follow this new structure.

- This large PR is mostly file moving, reconfiguring, and setting up — no major code behavior changes.

ThanhDatVu111 committed:
refactor: split project into frontend and backend, setup nativewind and fix Metro/Babel config

ThanhDatVu111 committed:
cleanup: remove obsolete old structure after frontend-backend split

ThanhDatVu111 committed:
add prettier to format all the files again

ThanhDatVu111 committed:
chore(monorepo): restructure package.json and tsconfig for frontend, backend, and root

- Added individual package.json files for frontend and backend
- Introduced tsconfig.json in backend to support TypeScript
- Moved shared dev dependencies (prettier, eslint) to root
- Enabled workspace-style development for cleaner dependency management

ThanhDatVu111 committed:
feat: connect frontend onboarding to backend API and Supabase

- Replaced direct Supabase insert in frontend with a POST request to Express backend
- Created  route to handle new user creation via Supabase
- Added Supabase client in backend with  config
- Confirmed working data flow from frontend ➝ backend ➝ Supabase
- Verified Supabase write via Supabase dashboard

ThanhDatVu111 committed:
feat: connect frontend onboarding to backend API and Supabase
- Replaced direct Supabase insert in frontend with a POST request to Express backend
- Created  route to handle new user creation via Supabase
- Added Supabase client in backend with  config
- Confirmed working data flow from frontend ➝ backend ➝ Supabase
- Verified Supabase write via Supabase dashboard

yourvivian commented on backend/tsconfig.json:
```diff
@@ -0,0 +1,14 @@
+    "skipLibCheck": true, // Skip type checking for node_modules
+    "forceConsistentCasingInFileNames": true
+  },
+  "include": ["src/**/*"],
```
```suggestion
  "include": ["**/*.ts", "**/*.tsx"],
```
to include all typescript files, not just in src

ThanhDatVu111 replied:
done

yourvivian commented on backend/src/routes/userRoutes.ts:
```diff
@@ -0,0 +1,33 @@
+  res.status(201).json({ message: '✅ User created', data });
+});
+
+export default router;
```
Let's split the logic into layers: controller, service, model, and routes to have a cleaner structure + separation of concerns
```
backend/
  src/
    controllers/
      userController.ts
    models/
      userModel.ts
    services/
      userService.ts
    routes/
      userRoutes.ts
    db.ts
    server.ts
```
- The Model layer handles database interactions. Move the Supabase query logic here.
- The Service layer calls the model and handles any additional processing.
- The Controller layer handles HTTP requests and responses
- The Routes layer connects the controller to the route.

yourvivian replied:
You can read more by searching about mvc pattern: https://blog.logrocket.com/building-structuring-node-js-mvc-application/

ThanhDatVu111 replied:
done

aanh1009 commented on frontend/tailwind.config.js:
```diff
@@ -1,20 +1,20 @@
-  content: ["./app/**/*.{js,jsx,ts,tsx}", "./components/**/*.{js,jsx,ts,tsx}"],
-  presets: [require("nativewind/preset")],
+  content: ['./src/app/**/*.{js,jsx,ts,tsx}'],
+  presets: [require('nativewind/preset')],
```
Right now, the way `frontend` is being structured is: we have a subfolder `src` and another subfolder `app` nested directly inside it. Subfolder `app` contains all other subfolders: auth, hooks, onboard, assets, components, constants. My suggestion is:
- For reusable components, nest them in `src/`: hooks, assets, components, constants
- For non-reusable components, nest them in `src/app/`: auth, onboard.

So correspondingly, the code here should be: `content: ['./src/**/*.{js,jsx,ts,tsx}'],`

ThanhDatVu111 replied:
done

yourvivian replied:
@aanh1009 
To clarify, it’s not simply about whether something is reusable or not. In Expo Router, `src/app/` is for route-based screens. Shared utilities like hooks/, components/, or constants/ that are reused across routes should live in `src/` (will not have a URL, and they cannot be the target of a navigation action, [ref](https://docs.expo.dev/router/basics/core-concepts/)). The split is about routing scope, not reusability.

tranguv commented on frontend/src/app/components/AuthLayout.tsx:
```diff
@@ -20,19 +20,11 @@ interface AuthLayoutProps {
+export default function AuthLayout({ children, activeTab, onTabChange }: AuthLayoutProps) {
   return (
     <View className="flex-1 items-center justify-center bg-primary">
       // logo
```
This is not valid JSX comment, should change to {/* logo */}
This will cause the '_Text strings must be rendered within a <Text> component_' error

ThanhDatVu111 replied:
Thank you for spotting this out for us

tranguv commented on frontend/src/app/components/AuthLayout.tsx:
```diff
@@ -20,19 +20,11 @@ interface AuthLayoutProps {
-        resizeMode="contain"
-      />
+      <Image source={images.logo} className="w-full max-h-28" resizeMode="contain" />
       // tagline
```
This is not valid JSX comment

tranguv reviewed (changes requested):
LGTM

hoangr1010 commented on backend/src/db.ts:
```diff
@@ -0,0 +1,8 @@
+
+dotenv.config();
+
+const supabase = createClient(process.env.SUPABASE_URL!, process.env.SUPABASE_ANON_KEY!);
```
why don't you leave this in supabase folder? if we already separate database to a single folder, should we totally separate it?

ThanhDatVu111 replied:
Moved!

ThanhDatVu111 committed:
chore(frontend): restructure project layout to separate reusable and page‑specific code

- Move reusable modules (hooks, assets, components, constants) into top‑level  directory
- Limit  to feature‑specific folders (, , etc.)
- Update Tailwind config “content” paths to: content: ['./src/**/*.{js,jsx,ts,tsx}'],

ThanhDatVu111 committed:
refactor(backend): split user logic into routes, controllers, services, and models

- Introduce layered structure under :
  -  for Express route definitions
  -  for HTTP request/response handling
  -  for business logic orchestration
  -  for Supabase database interactions

hoangr1010 commented on backend/src/server.ts:
```diff
@@ -0,0 +1,36 @@
+app.use('/user', UserRouter);
+
+// ✅ Start server
+const startServer = async () => {
```
this is fine, but for catching error we should have solid try catch for both database and server
```  try {
    console.log('🔎 Checking Supabase connectivity…');
    const { data, error } = await supabase
      .from('users')
      .select('*')
      .limit(1);

    if (error) {
      // Turn Supabase errors into thrown exceptions
      throw new Error(`Supabase health check failed: ${error.message}`);
    }

    // If we get here, connectivity is good
    app.listen(PORT, () => {
      console.log(`✅ Server running at ${LOCAL_HOST_URL}${PORT}`);
    });

  } catch (err) {
    // Catches both Supabase errors and unexpected runtime errors
    console.error('🚨 Failed to start server:', err.message || err);
    // Exit with failure code so process managers (PM2, Docker, etc.) know something went wrong
    process.exit(1);
  }
```

ThanhDatVu111 replied:
Yess sirrrr!

hoangr1010 commented on frontend/src/app/(onboard)/onboardingFlow.tsx:
```diff
@@ -191,37 +185,35 @@ const OnboardingFlow = () => {
+      const result = await response.json();
+
+      if (!response.ok) {
+        throw new Error(result.error || 'Something went wrong');
```
Suggestion: make a folder `frontend/apis` where makes api request instead of leaves directly in the components file, you can make subfolder/file where indicate each page, then create request in it.
Also consider to [use axios](https://rapidapi.com/guides/custom-axios-instance), where you can config the request and avoid duplicate code, such as config header, baseURL,... you can explore more in [the docs](https://axios-http.com/docs/api_intro)

ThanhDatVu111 replied:
this api folder has to be inside src right? , frontend/src/apis?

hoangr1010 reviewed (approved):
look good overall, great job!! left few suggestions.

ThanhDatVu111 committed:
separate api call logic from onboardingflow.ts

ThanhDatVu111 committed:
move db into supabase and add minor changes for other files , added env.example too

ThanhDatVu111 committed:
added quick instruction inside readme

MC2114 reviewed (approved):
Great work from anh Dat! The code structure is very clean (and chi Trang's suggestions was very helpful on how we should split the components / screens going onwards). I have tested this on my machine and it works just fine; the routing is correct, the codebase is well-structured and much more navigable now.

ThanhDatVu111 committed:
added screens and containers folders
